core/basic: argument invariance for all Atom classes - class

ThePauliPrinciple opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
The following identities now hold for `Atom` classes:
```python
assert obj.func == AtomSubClass
assert AtomSubClass == obj.func
assert hash(AtomSubClass) == hash(obj.func)
assert {AtomSubClass: True}[obj.func]
assert {obj.func: True}[AtomSubClass]
assert obj.func() == obj
assert obj.func(*obj.args) == obj
assert str(obj.func) == str(AtomSubClass)
```
These seem to all be needed since several parts of SymPy
assumes that `obj.func` returns `obj.__class__`, although it
is not needed for every class.

Several classes still behave as previously, these are:

`Quantity`
`BaseVector`
`BaseDyadic`
`BaseScalar`

It seems that these classes need a redesign rather than
a fix to their argument invariance, hence the previous
behaviour was maintained.

#### Other comments
It might be the case that some "solutions" provided to issues I encountered are actually not needed as the issues were only introduced due to changes I made (and reverted).

```python
@property
def func(self):
    return lambda: self
```
Has several intertwined issues and I expanded from that. In particular:
* printing of `obj.func`
* `Singleton`s and their "create once" behaviour
* comparisons of `obj.func == obj.class`
* use of `obj.__class__` hash for comparison with `obj.func` hash (e.g. in `dict` lookups)

were use cases in SymPy that were assumed to work as if `obj.func` always returns
`obj.__class__`.

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* core
  * Atom classes now correctly implement the invariance `obj.func(*obj.args)`
<!-- END RELEASE NOTES -->

ThePauliPrinciple commented:
Looks like this leaves behind some horrible slowdowns, I have two things that could improve performance that came immediately to mind:
* caching `obj.func` 
* overriding `ManagedProperties.__init__` 

I will try these seperately, so their individual effect becomes clear from the benchmark.

ThePauliPrinciple commented:
Current benchmark results appear quite favourable:

<details>
<summary> full benchmark </summary>

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/paul/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/paul/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>
<!--table
	{mso-displayed-decimal-separator:"\,";
	mso-displayed-thousand-separator:"\.";}
@page
	{margin:.75in .7in .75in .7in;
	mso-header-margin:.3in;
	mso-footer-margin:.3in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:11.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:Calibri, sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
-->
</style>
</head>

<body link="#0563C1" vlink="#954F72">



n/a | n/a | n/a | solve.TimeSolveLinSys.time_solve_lin_sys
-- | -- | -- | --
n/a | n/a | n/a | solve.TimeSolveLinSys.time_to_expr_eqs
n/a | n/a | n/a | solve.TimeSolveLinSys.time_verify_sol
n/a | n/a | n/a | solve.TimeSolveLinSys10x8.time_solve_lin_sys
n/a | n/a | n/a | solve.TimeSolveLinSys10x8.time_to_expr_eqs
n/a | n/a | n/a | solve.TimeSolveLinSys10x8.time_verify_sol
n/a | n/a | n/a | solve.TimeSolveLinSys165x165.time_solve_lin_sys
n/a | n/a | n/a | solve.TimeSolveLinSys165x165.time_to_expr_eqs
n/a | n/a | n/a | solve.TimeSolveLinSys165x165.time_verify_sol
894±20μs | 993±90μs | 1.11 | cse.TimeCSE.time_cse((2,
4.26±0.1μs | 4.33±0.4μs | 1.02 | core.expand.TimeExpand.time_expand_modulus
894±50μs | 905±40μs | 1.01 | cse.TimeCSE.time_cse((2,
54.2M | 54.2M | 1.00 | large_exprs.TimeLargeExpressionOperations.peakmem_jacobian_wrt_functions
53.3M | 53.3M | 1.00 | large_exprs.TimeLargeExpressionOperations.peakmem_jacobian_wrt_symbols
53.3M | 53.3M | 1.00 | large_exprs.TimeLargeExpressionOperations.peakmem_subs
2.22±0.07μs | 2.20±0.07μs | 0.99 | core.arit.Time_Arit.time_Add_x1
813±20μs | 802±40μs | 0.99 | cse.TimeCSE.time_cse((3,
38.0±0.6ms | 37.4±1ms | 0.99 | integrate.TimeIntegrationRisch01.time_doit_risch(10)
29.1±1ms | 28.4±2ms | 0.98 | integrate.TimeIntegrationRisch01.time_doit_risch(1)
821±30μs | 802±50μs | 0.98 | solve.TimeMatrixSolvePyDySlow.time_cholesky_solve(1)
12.2±0.5μs | 11.7±0.7μs | 0.96 | core.expand.TimeExpand.time_expand_frac
1.52±0.05ms | 1.46±0.1ms | 0.96 | solve.TimeMatrixSolvePyDyFast.time_LUsolve(3)
672±20μs | 648±20μs | 0.96 | solve.TimeRationalSystem.time_linsolve(5)
291±5ms | 276±7ms | 0.95 | integrate.TimeIntegrationRisch01.time_doit(100)
142±2ms | 136±5ms | 0.95 | logic.LogicSuite.time_dpll2
950±20μs | 900±30μs | 0.95 | solve.TimeMatrixSolvePyDySlow.time_LDLsolve(1)
2.61±0.1ms | 2.47±0.06ms | 0.95 | solve.TimeRationalSystem.time_linsolve(10)
8.10±0.2μs | 7.65±0.3μs | 0.94 | core.expand.TimeExpand.time_expand_no_log
5.30±0.07ms | 4.98±0.2ms | 0.94 | polygon.PolygonAttributes.time_create
682±30μs | 642±30μs | 0.94 | solve.TimeMatrixSolvePyDyFast.time_LUsolve(1)
346±10μs | 324±30μs | 0.94 | solve.TimeRationalSystem.time_linsolve(3)
2.26±0.1μs | 2.10±0.1μs | 0.93 | core.arit.Time_Arit.time_Add_1x
17.4±0.8μs | 16.2±0.5μs | 0.93 | core.arit.Time_Arit.time_Add_x05
1.36±0.09μs | 1.26±0.1μs | 0.93 | core.arit.Time_Arit.time_neg
477±10ns | 442±20ns | 0.93 | core.expand.TimeExpand.time_expand_nothing_todo
758±40μs | 703±50μs | 0.93 | cse.TimeCSE.time_cse((2,
55.5±0.7ms | 51.6±2ms | 0.93 | integrate.TimeIntegration01.time_doit_meijerg
2.29±0.07μs | 2.14±0.08μs | 0.93 | integrate.TimeIntegrationRisch02.time_doit(1)
3.43±0.5ms | 3.19±0.08ms | 0.93 | solve.TimeSparseSystem.time_linear_eq_to_matrix(20)
557±20ns | 512±30ns | 0.92 | core.arit.Time_Arit.time_pow100
8.82±0.5μs | 8.15±0.6μs | 0.92 | core.expand.TimeExpand.time_expand_negative_integers_powers
16.1±0.5μs | 14.8±1μs | 0.92 | core.expand.TimeExpand.time_expnad_radicals
126±2ms | 115±3ms | 0.92 | integrate.TimeIntegration01.time_doit
4.59±0.3μs | 4.20±0.07μs | 0.92 | matrices.TimeMatrixExpression.time_Add
187±2ms | 172±9ms | 0.92 | solve.TimeMatrixSolve.time_solve_sym('ADJ')
530±10μs | 481±30μs | 0.91 | core.arit.Time_Arit.time_mod_pow
803±30μs | 730±80μs | 0.91 | cse.TimeCSE.time_cse((4,
304±9ms | 276±10ms | 0.91 | integrate.TimeIntegrationRisch01.time_doit_risch(100)
97.2±2ms | 88.3±3ms | 0.91 | matrices.TimeMatrixPower.time_Case3
844±20ms | 766±20ms | 0.91 | polygon.PolygonFirstMoment.time_bench01
105±4μs | 95.2±3μs | 0.91 | solve.TimeMatrixOperations.time_det_bareiss(3,
56.2±2ms | 51.4±2ms | 0.91 | solve.TimeMatrixSolvePyDySlow.time_linsolve(1)
4.54±0.1ms | 4.11±0.08ms | 0.91 | solve.TimeRationalSystemSymbol.time_solve(3)
14.8±0.9ms | 13.5±0.9ms | 0.91 | solve.TimeSolveSparsePolySystem.time_solve(1)
6.78±0.4ms | 6.16±0.4ms | 0.91 | solve.TimeSparseSystem.time_linear_eq_to_matrix(30)
186±2ms | 167±4ms | 0.90 | baseline.TimeBaseline.time_loop10000000
28.2±0.8ms | 25.4±2ms | 0.90 | integrate.TimeIntegrationRisch01.time_doit(1)
18.4±0.5ms | 16.5±1ms | 0.90 | large_exprs.TimeLargeExpressionOperations.time_cse
9.66±0.3ms | 8.70±0.3ms | 0.90 | logic.LogicSuite.time_load_file
264±20μs | 238±10μs | 0.90 | matrices.TimeMatrixExpression.time_MatMul
16.2±0.7ms | 14.6±0.4ms | 0.90 | matrices.TimeMatrixExpression.time_MatMul_doit
677±30μs | 611±40μs | 0.90 | solve.TimeMatrixOperations.time_rank(4,
1.03±0.03ms | 933±40μs | 0.90 | solve.TimeMatrixSolve.time_solve_sym('LU')
9.53±0.3ms | 8.62±0.7ms | 0.90 | solve.TimeRationalSystem.time_solve(5)
28.8±0.8ms | 26.0±0.5ms | 0.90 | solve.TimeSolveSparsePolySystem.time_solve(2)
138±7ms | 124±6ms | 0.90 | sum.TimeSum.time_doit
1.33±0.07μs | 1.18±0.06μs | 0.89 | core.arit.Time_Arit.time_Add_xy
137±5ms | 121±6ms | 0.89 | dsolve.TimeDsolve01.time_dsolve
100±10ms | 89.8±20ms | 0.89 | polygon.PolygonDistance.time_bench01
3.90±0.1ms | 3.46±0.1ms | 0.89 | polygon.PolygonSecondMoment.time_bench01
5.83±0.1ms | 5.20±0.5ms | 0.89 | polygon.PolygonSectionModulus.time_bench01
188±4μs | 168±3μs | 0.89 | solve.TimeMatrixOperations.time_rref(3,
250±8ms | 223±4ms | 0.89 | solve.TimeMatrixSolve.time_solve('ADJ')
1.01±0.03ms | 905±30μs | 0.89 | solve.TimeMatrixSolvePyDyFast.time_LUsolve(2)
56.6±1ms | 50.3±2ms | 0.89 | solve.TimeMatrixSolvePyDySlow.time_solve(1)
30.3±1ms | 26.8±1ms | 0.89 | solve.TimeRationalSystem.time_solve(10)
4.23±0.1ms | 3.78±0.2ms | 0.89 | solve.TimeRationalSystem.time_solve(3)
28.1±0.9ms | 25.0±0.6ms | 0.89 | solve.TimeSolve01.time_solve_nocheck
17.9±0.6ms | 15.9±0.9ms | 0.89 | solve.TimeSparseSystem.time_linsolve_Aaug(10)
95.6±3ms | 84.8±4ms | 0.89 | solve.TimeSparseSystem.time_linsolve_Aaug(30)
1.31±0.06μs | 1.15±0.02μs | 0.88 | baseline.TimeBasicGCD.time_gcd
866±20ns | 763±50ns | 0.88 | core.arit.Time_Arit.time_Add_xyz
550±10ns | 486±10ns | 0.88 | core.arit.Time_Arit.time_pow2
21.1±0.6μs | 18.6±0.5μs | 0.88 | core.arit.Time_Arit.time_pow_im
9.64±0.3μs | 8.50±0.7μs | 0.88 | core.arit.Time_assumption.time_Add_Mul_is_integer
39.4±0.8ms | 34.6±2ms | 0.88 | integrate.TimeIntegrationRisch01.time_doit(10)
94.0±2ms | 82.8±2ms | 0.88 | integrate.TimeIntegrationRisch02.time_doit(10)
193±9μs | 169±7μs | 0.88 | polygon.PolygonAttributes.time_area
24.1±1μs | 21.3±1μs | 0.88 | polys.TimePolyManyGens.time_is_linear(10)
104±5μs | 91.3±4μs | 0.88 | solve.TimeMatrixOperations.time_det(3,
59.5±3ms | 52.6±1ms | 0.88 | solve.TimeMatrixOperations.time_rank(4,
3.17±0.3μs | 2.76±0.2μs | 0.87 | core.arit.Time_Arit.time_Div_2y
7.73±0.2μs | 6.71±0.7μs | 0.87 | core.expand.TimeExpand.time_expand_complex_number_1
4.45±0.09s | 3.87±0.2s | 0.87 | integrate.TimeIntegrationRisch02.time_doit(100)
1.13±0.01s | 983±10ms | 0.87 | large_exprs.TimeLargeExpressionOperations.time_jacobian_wrt_functions
3.04±0.01s | 2.63±0.02s | 0.87 | logic.LogicSuite.time_dpll
12.3±0.5ms | 10.7±0.7ms | 0.87 | matrices.TimeMatrixPower.time_Case1
107±5ms | 93.2±1ms | 0.87 | matrices.TimeMatrixPower.time_Case2
752±40μs | 654±10μs | 0.87 | polygon.PolygonAttributes.time_perimeter
1.11±0.04s | 971±20ms | 0.87 | polygon.PolygonCutSection.time_bench01
240±10μs | 208±4μs | 0.87 | solve.TimeMatrixArithmetic.time_dense_multiply(4,
38.4±1μs | 33.4±1μs | 0.87 | solve.TimeMatrixOperations.time_det_berkowitz(3,
59.3±1ms | 51.5±2ms | 0.87 | solve.TimeMatrixOperations.time_rank(4,
29.7±0.2ms | 25.8±0.9ms | 0.87 | solve.TimeMatrixOperations.time_rref(3,
48.7±3ms | 42.6±4ms | 0.87 | solve.TimeMatrixPositiveDefinite.time_A_sym_is_positive_definite
678±10ms | 590±10ms | 0.87 | solve.TimeMatrixSolve.time_solve('GE')
999±30μs | 868±20μs | 0.87 | solve.TimeMatrixSolve.time_solve('LU')
1.03±0.04ms | 902±40μs | 0.87 | solve.TimeMatrixSolve2.time_lusolve
1.35±0.05ms | 1.18±0.06ms | 0.87 | solve.TimeRationalSystemSymbol.time_solve(1)
29.7±0.6ms | 25.8±0.8ms | 0.87 | solve.TimeSolve01.time_solve
46.2±3ms | 40.3±1ms | 0.87 | solve.TimeSolveLinSys189x49.time_to_expr_eqs
64.2±3ms | 56.1±2ms | 0.87 | solve.TimeSparseSystem.time_solve(30)
869±10ns | 743±50ns | 0.86 | core.arit.Time_Arit.time_Mul_xyz
28.4±1μs | 24.5±0.6μs | 0.86 | core.arit.Time_assumption.time_ncmul
11.3±0.3μs | 9.74±0.2μs | 0.86 | core.expand.TimeExpand.time_expand_complex_number_2
32.7±1μs | 28.2±1μs | 0.86 | core.expand.TimeExpand.time_expand_log
12.9±0.4μs | 11.1±0.6μs | 0.86 | core.expand.TimeExpand.time_expand_power_base
766±20μs | 659±30μs | 0.86 | cse.TimeCSE.time_cse((2,
96.1±4ms | 82.6±0.9ms | 0.86 | integrate.TimeIntegrationRisch02.time_doit_risch(10)
27.7±0.8μs | 23.9±0.9μs | 0.86 | matrices.TimeDiagonalEigenvals.time_eigenvals
17.5±0.3ms | 15.1±0.5ms | 0.86 | matrices.TimeMatrixExpression.time_MatAdd_doit
7.96±0.3ms | 6.81±0.2ms | 0.86 | physics.mechanics.kane.KanesMethodMassSpringDamper.time_kanesmethod_mass_spring_damper
574±10μs | 496±10μs | 0.86 | solve.TimeMatrixArithmetic.time_dense_multiply(6,
577±10ms | 498±10ms | 0.86 | solve.TimeMatrixSolve.time_solve_sym('GE')
1.23±0.05ms | 1.06±0.02ms | 0.86 | solve.TimeRationalSystem.time_solve(1)
45.2±0.5ms | 38.9±2ms | 0.86 | solve.TimeSolveSparsePolySystem.time_solve(3)
93.2±1ms | 80.3±4ms | 0.86 | solve.TimeSparseSystem.time_linsolve_Ab(30)
11.8±0.4μs | 10.0±0.7μs | 0.85 | core.expand.TimeExpand.time_expand_no_multinomial
12.3±0.4μs | 10.4±0.1μs | 0.85 | core.expand.TimeExpand.time_expand_non_commutative
7.18±0.2ms | 6.07±0.4ms | 0.85 | large_exprs.TimeLargeExpressionOperations.time_free_symbols
341±3ms | 290±2ms | 0.85 | polygon.PolygonArbitraryPoint.time_bench01
4.00±0.2ms | 3.40±0.1ms | 0.85 | polygon.PolygonPolarSecondMoment.time_bench01
1.88±0.1μs | 1.60±0.1μs | 0.85 | polys.TimePolyManyGens.time_is_linear(1)
128±4μs | 109±3μs | 0.85 | solve.TimeMatrixOperations.time_det_bareiss(3,
759±20μs | 648±40μs | 0.85 | solve.TimeMatrixOperations.time_det_bareiss(4,
5.49±0.09ms | 4.66±0.1ms | 0.85 | solve.TimeMatrixOperations.time_det_bareiss(4,
1.86±0.07ms | 1.58±0.06ms | 0.85 | solve.TimeMatrixOperations.time_det_berkowitz(4,
54.0±0.5ms | 45.9±2ms | 0.85 | solve.TimeMatrixOperations.time_rank(3,
27.6±0.7ms | 23.5±0.6ms | 0.85 | solve.TimeMatrixOperations.time_rank(3,
346±6μs | 292±20μs | 0.85 | solve.TimeMatrixOperations.time_rref(4,
47.5±2ms | 40.6±0.7ms | 0.85 | solve.TimeMatrixPositiveDefinite.time_A_is_positive_definite
1.45±0.06ms | 1.23±0.06ms | 0.85 | solve.TimeMatrixSolve2.time_LDLsolve
138±6μs | 117±8μs | 0.85 | solve.TimeRationalSystem.time_linsolve(1)
504±10μs | 428±10μs | 0.85 | solve.TimeRationalSystemSymbol.time_linsolve(1)
1.48±0.01ms | 1.25±0.01ms | 0.85 | solve.TimeRationalSystemSymbol.time_linsolve(3)
1.31±0.02ms | 1.11±0.04ms | 0.85 | solve.TimeSparseSystem.time_linear_eq_to_matrix(10)
50.0±4ms | 42.2±0.3ms | 0.85 | solve.TimeSparseSystem.time_linsolve_Aaug(20)
31.1±1μs | 26.1±0.8μs | 0.84 | core.arit.Time_assumption.time_ncpow
33.5±0.5ms | 28.2±0.5ms | 0.84 | lambdify.TimeLambdifyCreation.time_lambdify_create(['numpy'])
20.2±2ms | 17.1±0.5ms | 0.84 | physics.mechanics.lagrange.LagrangesMethodMassSpringDamper.time_lagrangesmethod_mass_spring_damper
708±20μs | 592±10μs | 0.84 | polygon.PolygonAttributes.time_sides
156±6ms | 132±1ms | 0.84 | polygon.PolygonEnclosesPoint.time_bench01
45.1±1ms | 38.0±1ms | 0.84 | polygon.PolygonIsConvex.time_bench01
113±4μs | 95.1±3μs | 0.84 | polys.TimePolyManyGens.time_create_poly(10)
14.4±0.5ms | 12.2±0.2ms | 0.84 | polys.TimePolyManyGens.time_is_linear(500)
53.3±4μs | 45.0±2μs | 0.84 | solve.TimeMatrixArithmetic.time_dense_add(10,
30.9±2μs | 25.9±0.6μs | 0.84 | solve.TimeMatrixArithmetic.time_dense_multiply(4,
129±5μs | 108±1μs | 0.84 | solve.TimeMatrixOperations.time_det(3,
808±20μs | 682±40μs | 0.84 | solve.TimeMatrixOperations.time_det(4,
493±40μs | 417±10μs | 0.84 | solve.TimeMatrixOperations.time_rank(3,
9.94±0.3ms | 8.38±0.3ms | 0.84 | solve.TimeRationalSystemSymbol.time_solve(5)
39.5±0.5ms | 33.0±2ms | 0.84 | solve.TimeSolveLinSys189x49.time_solve_lin_sys
95.6±2ms | 80.3±2ms | 0.84 | solve.TimeSolveLinSys189x49.time_verify_sol
338±10ns | 281±10ns | 0.83 | baseline.TimeBaseline.time_loop10
70.4±0.6ns | 58.6±2ns | 0.83 | baseline.TimeBaseline.time_noop
2.20±0.09μs | 1.84±0.07μs | 0.83 | core.arit.Time_Arit.time_Div_xy
1.36±0.05μs | 1.13±0.02μs | 0.83 | core.arit.Time_Arit.time_Mul_xy
10.0±0.3μs | 8.32±0.1μs | 0.83 | core.expand.TimeExpand.time_expand_arit
3.65±0.2μs | 3.01±0.09μs | 0.83 | integrate.TimeIntegrationRisch02.time_doit_risch(1)
4.42±0.05s | 3.66±0.03s | 0.83 | integrate.TimeIntegrationRisch02.time_doit_risch(100)
2.52±0.03s | 2.08±0.03s | 0.83 | integrate.TimeIntegrationRisch03.time_doit_risch(100)
30.0±1ms | 24.9±1ms | 0.83 | large_exprs.TimeLargeExpressionOperations.time_jacobian_wrt_symbols
1.18±0.02s | 977±6ms | 0.83 | large_exprs.TimeLargeExpressionOperations.time_manual_jacobian_wrt_functions
39.8±2μs | 33.0±2μs | 0.83 | solve.TimeMatrixOperations.time_det(3,
5.42±0.06ms | 4.49±0.1ms | 0.83 | solve.TimeMatrixOperations.time_det(4,
1.21±0.04ms | 1.00±0.03ms | 0.83 | solve.TimeMatrixOperations.time_det_berkowitz(4,
22.0±0.4ms | 18.2±1ms | 0.83 | solve.TimeSparseSystem.time_solve(10)
57.5±1μs | 47.4±1μs | 0.82 | lambdify.TimeLambdifyEvaluate.time_lambdify_evaluate(['numpy'])
52.7±2μs | 43.0±0.9μs | 0.82 | matrices.TimeMatrixGetItem.time_MutableSparseMatrix_getitem
1.10±0.06ms | 898±20μs | 0.82 | polygon.PolygonAttributes.time_centroid
65.8±2μs | 54.1±2μs | 0.82 | solve.TimeMatrixArithmetic.time_dense_add(6,
1.75±0.05ms | 1.45±0.01ms | 0.82 | solve.TimeMatrixOperations.time_det_berkowitz(4,
57.4±3ms | 47.0±1ms | 0.82 | solve.TimeMatrixOperations.time_rref(3,
76.6±2ms | 63.2±1ms | 0.82 | solve.TimeMatrixOperations.time_rref(4,
86.2±3ms | 71.0±2ms | 0.82 | solve.TimeMatrixOperations.time_rref(4,
3.48±0.1ms | 2.87±0.02ms | 0.82 | solve.TimeRationalSystemSymbol.time_linsolve(5)
1.51±0.05ms | 1.24±0.02ms | 0.82 | solve.TimeSparseSystem.time_linsolve_eqs(20)
2.19±0.05ms | 1.80±0.07ms | 0.82 | solve.TimeSparseSystem.time_linsolve_eqs(30)
6.95±0.3μs | 5.65±0.3μs | 0.81 | core.arit.Time_assumption.time_Add_Mul_is_finite
33.5±0.8ms | 27.2±0.3ms | 0.81 | lambdify.TimeLambdifyCreation.time_lambdify_create(['math'])
70.6±4μs | 57.0±2μs | 0.81 | polys.TimePolyManyGens.time_create_poly(1)
926±40μs | 750±20μs | 0.81 | polys.TimePolyManyGens.time_create_poly(100)
15.4±0.2ms | 12.5±0.2ms | 0.81 | polys.TimePolyManyGens.time_create_poly(500)
694±40μs | 564±10μs | 0.81 | polys.TimePolyManyGens.time_is_linear(100)
1.50±0.1ms | 1.22±0.04ms | 0.81 | solve.TimeMatrixOperations.time_det_bareiss(4,
127±9μs | 102±3μs | 0.81 | solve.TimeMatrixOperations.time_det_berkowitz(3,
103±2μs | 84.0±0.8μs | 0.81 | solve.TimeMatrixOperations.time_det_berkowitz(3,
868±30μs | 699±10μs | 0.81 | solve.TimeSparseSystem.time_linsolve_eqs(10)
220±9ms | 177±4ms | 0.80 | integrate.TimeIntegrationRisch03.time_doit_risch(10)
44.9±1ms | 35.8±2ms | 0.80 | large_exprs.TimeLargeExpressionOperations.time_count_ops
27.4±1μs | 22.0±0.3μs | 0.80 | matrices.TimeMatrixExpression.time_MatAdd
5.07±0.4μs | 4.06±0.07μs | 0.80 | matrices.TimeMatrixExpression.time_Mul
51.9±2μs | 41.7±0.6μs | 0.80 | matrices.TimeMatrixGetItem.time_ImmutableDenseMatrix_getitem
30.3±1μs | 24.4±0.4μs | 0.80 | solve.TimeMatrixArithmetic.time_dense_add(3,
24.3±1μs | 19.5±0.5μs | 0.80 | solve.TimeMatrixArithmetic.time_dense_multiply(3,
134±8μs | 107±3μs | 0.80 | solve.TimeMatrixArithmetic.time_dense_multiply(3,
1.47±0.05ms | 1.17±0.01ms | 0.80 | solve.TimeMatrixOperations.time_det(4,
1.34±0.07ms | 1.07±0.03ms | 0.80 | solve.TimeMatrixSolve2.time_cholesky_solve
659±60μs | 518±20μs | 0.79 | integrate.TimeIntegrationRisch03.time_doit_risch(1)
305±9ms | 242±5ms | 0.79 | large_exprs.TimeLargeExpressionOperations.time_subs
141±3μs | 111±3μs | 0.79 | solve.TimeMatrixArithmetic.time_dense_add(10,
43.9±0.9ms | 34.7±0.3ms | 0.79 | solve.TimeSparseSystem.time_solve(20)
221±2ms | 172±2ms | 0.78 | integrate.TimeIntegrationRisch03.time_doit(10)
8.23±0.5μs | 6.42±0.2μs | 0.78 | lambdify.TimeLambdifyEvaluate.time_lambdify_evaluate(['math'])
53.1±4μs | 41.3±0.9μs | 0.78 | matrices.TimeMatrixGetItem.time_ImmutableSparseMatrix_getitem
52.1±3μs | 40.5±0.8μs | 0.78 | matrices.TimeMatrixGetItem.time_MutableDenseMatrix_getitem
69.2±2μs | 53.8±0.3μs | 0.78 | solve.TimeMatrixArithmetic.time_dense_multiply(6,
39.8±1μs | 31.2±0.4μs | 0.78 | solve.TimeMatrixOperations.time_det_bareiss(3,
34.8±2μs | 26.9±0.5μs | 0.77 | integrate.TimeIntegrationRisch03.time_doit(1)
2.71±0.03s | 2.09±0.02s | 0.77 | integrate.TimeIntegrationRisch03.time_doit(100)
19.1±2μs | 14.8±0.4μs | 0.77 | solve.TimeMatrixArithmetic.time_dense_add(3,
19.4±1ms | 15.0±0.5ms | 0.77 | solve.TimeSparseSystem.time_linsolve_Ab(10)
31.6±0.6μs | 24.0±0.6μs | 0.76 | solve.TimeMatrixArithmetic.time_dense_add(6,
249±8μs | 189±5μs | 0.76 | solve.TimeMatrixArithmetic.time_dense_multiply(10,
51.6±2ms | 39.1±1ms | 0.76 | solve.TimeSparseSystem.time_linsolve_Ab(20)
43.1±2μs | 31.9±0.9μs | 0.74 | solve.TimeMatrixArithmetic.time_dense_add(4,
22.8±0.6μs | 16.5±0.2μs | 0.72 | solve.TimeMatrixArithmetic.time_dense_add(4,
1.91±0.3ms | 1.31±0.06ms | 0.68 | solve.TimeMatrixArithmetic.time_dense_multiply(10,



</body>

</html>
</details>
Am I misunderstanding these? It seems suspecious.

smichr commented on sympy/core/singleton.py:
```diff
@@ -175,6 +175,9 @@ def __init__(cls, *args, **kwargs):
         cls.__getstate__ = lambda obj: None
+        def func(self):
+            return self.__class__
+        cls.func = property(func)
```
Could this be `cls.func = property(lambda _: _.__class__)`?

ThePauliPrinciple replied:
I think so. Do you think it would increase performance? Or is this considered to be better style? (possibly both)

smichr replied:
not sure about performance; just styl considering the surrounding code

smichr commented on sympy/core/basic.py:
```diff
@@ -1943,6 +1961,16 @@ def _sorted_args(self):
+    @property
+    def func(self):
+        return self._func_hook()
+    @cacheit
```
blank line between these two is needed

smichr commented on sympy/core/basic.py:
```diff
@@ -1896,6 +1896,24 @@ def could_extract_minus_sign(self):
+    def __init__(self, name, bases, dct):
+        pass
+    def __subclasscheck__(cls, subclass):
+        if cls == subclass: return True
```
Please make the one-liners `if foo: do` into two lines with indentation on the second and leave a blank line bewteen functions.

smichr commented on sympy/core/tests/test_args.py:
```diff
@@ -98,7 +98,7 @@ def _test_args(obj):
     all_basic = all(isinstance(arg, Basic) for arg in obj.args)
     # Ideally obj.func(*obj.args) would always recreate the object, but for
     # now, we only require it for objects with non-empty .args
```
```suggestion
```

smichr commented:
This seems to be working in line with what we want the invariant to be, but after looking at the intelligent discussion on the issues page, I defer committal to someone with a more critical sense of the issue.

oscarbenjamin commented on sympy/core/basic.py:
```diff
@@ -1943,6 +1968,17 @@ def _sorted_args(self):
+
+    @cacheit
+    def _func_hook(self):
+        class FuncHook(self.__class__, metaclass = FuncHookMeta):
```
I'm not sure about this. Dynamically created classes are used in some other parts of SymPy e.g. `Function('x')` and always seem to end up causing problems somewhere.

ThePauliPrinciple replied:
The only reason I'm dynamically creating the class is because the result of `obj.func` needs to be able to pass an `issubclass` check (it is used at least in `sathandler.py`) and there is no way to pass an issubclass if the object you are checking is not a class object. Remember, class objects are not something special, they are just objects that have certain properties. (and the only way to get those properties is to create the object with `type` rather than `object`)

ThePauliPrinciple replied:
I can modify a `lambda` function to pass the hash checks (so they can be used as keys for dicts) and modify the `__eq__` behaviour, but `issubclass` can not be messed with unless you work with metaclasses.

If the `sathandler.py` can somehow be modified I'd be happy with that too, but it didn't seem obvious to me how that could be done.

ThePauliPrinciple replied:
Actually, I suppose it is possible, I could simply create a normal object and set its `__class__ = type` and `__mro__ = self.__class__.__mro__` as well as its `__eq__` and `__hash__` to what I want and it might work like that. Not sure if there is anything in python that would complain about it.

ThePauliPrinciple replied:
I can write it like `return FuncHookMeta('FuncHook', (self.__class__), {'__new__': lambda cls: self})` if that helps. Note also that the created class never creates an instance of it, it returns the original object when `__new__` is invoked, so it really just acts like a normal object that has properties such that it can be compared as if it were a class object

oscarbenjamin replied:
Why is that needed at all?

ThePauliPrinciple replied:
Because the result of `obj.func` (note here, without brackets) is used in `issubclass` in some parts of the code, which means that whatever `obj.func` gives has to be a class.

ThePauliPrinciple replied:
`obj.func` is also used as a key in several occasions, expecting to be the same as `obj.__class__` and finally it is also in some occassions compared with `==` to be equal to `obj.__class__`

oscarbenjamin replied:
I think that the places where `obj.func` is used as an alias for `obj.__class__` should just be changed to use `obj.__class__`. We should not expect `obj.func` to have any properties beyond callability. Also it's more efficient to access `obj.__class__` directly.

ThePauliPrinciple replied:
I can throw up a PR with 
```python
@property
def func(self):
    return lambda: self
```
To expose how big the issue is if that is appreciated

oscarbenjamin replied:
Yes, that would be helpful. My expectation is that there would be a large number of errors coming mostly from just a few places in the codebase so it would good to fix a few of the more common ones or otherwise there will be too much noise to see what's going on.

oscarbenjamin replied:
Actually it can't literally be that because that will break everything. Perhaps something like:
```python
@property
def func(self):
    return lambda *args, **kwargs: self.__class__(*args, **kwargs)
```

ThePauliPrinciple replied:
> Actually it can't literally be that because that will break everything. Perhaps something like:
> 
> ```python
> @property
> def func(self):
>     return lambda *args, **kwargs: self.__class__(*args, **kwargs)
> ```

How is this different from
```python
@property
def func(self):
    return self.__class__
```

oscarbenjamin replied:
The difference is that it isn't the type of the object so it can't be used in `isinstance` and that it won'r work properly with `__eq__` etc. We still need `func` to be callable and that the result of calling it produces the expected new object. Otherwise everything breaks.

oscarbenjamin replied:
Oh sorry, I missed that you only meant to make this change for Atom rather than for all of Basic. I was expecting to change this for all Basic so that anything that assumes `obj.func is obj.__class__` would break.

ThePauliPrinciple replied:
But isn't the whole point that `Atom`s can't be reconstructed from their arguments since they have none, but they might still contain information that would identify them (e.g. 
```python
from sympy import Symbol
Symbol('hi').args
```

oscarbenjamin replied:
Yes, but if we wanted to identify all places in the codebase that abuse `func` then we should change `func` on `Basic` (in a PR that doesn't directly get merged).

oscarbenjamin replied:
Like this PR: #20641. I don't want to merge the main change there (to `Basic.__new__`) but the test failures and the diff show the problems that need to be fixed so if we made it so that the tests pass there then most of the changes would be good ones that could then be merged.

oscarbenjamin replied:
I think we might be talking past each other...

ThePauliPrinciple replied:
Yes, I think so too.

oscarbenjamin commented on sympy/vector/dyadic.py:
```diff
@@ -225,6 +225,12 @@ def _sympystr(self, printer):
+
+    @property
+    def func(self):
+        return self.__class__
```
I can't remember any more. Why are these `func` methods needed?

ThePauliPrinciple replied:
In the other PR, you had suggested to put comments here explaining what is going on, I will put those here too. But basically, these classes do not follow the design pattern for `Basic`s to such a degree that we could not find a way to make them work without essentially rewriting them. (e.g. they are storing all sorts of weird non-`Basic` arguments)

oscarbenjamin commented on sympy/core/basic.py:
```diff
@@ -1944,6 +1944,42 @@ def could_extract_minus_sign(self):
+        return hash(self.obj.__class__)
+
+    def __repr__(self):
+        return repr(self.obj.__class__)
```
Probably best to print as something like `_FuncHead(self.obj.__class__)` or something else that just makes it clear that this isn't actually the type. It's a weird object so it should look like one.

asmeurer commented:
I agree that `func` need not be something that passes as the type of the object. If it could, we would not need it in the first place. What are the lines of code that are assuming this?

Are the benchmarks at https://github.com/sympy/sympy/pull/22522#issuecomment-973293473 still accurate. It seems like it still has some slowdowns, if they are to be trusted. Does just using a lambda instead of a class improve the performance?

ThePauliPrinciple commented:
The slowdowns are in "master vs previous release". It seems to be a problem in CI (some tests are randomly taking ~twice as long, which means that after every test of a new master push, different tests are twice as slow/twice as fast)

I would guess that using `lambda` instead of a `class` improves performance, but I would be surprised if that is often the performance critical part. Since the classes that have been changed can currently not be reconstructed with `func`, I don't think there are any timings available for this.

We had already previously made many commits that would allow using `lambda`, but recently an `isinstance(..., .func)` ended in a PR: https://github.com/sympy/sympy/pull/22841

While this is done for a class where `.func` does return a `type`, it just feels like we are asking for potential silent bugs if devs are still expecting `.func` to return a `type` (not to mention downstream). While I'm not exactly against changing such a design pattern, `func` is really at the core of how SymPy works and it seems to me that even among active developers there exists no such consensus.

This PR is not as invasive and instead it only affects parts of SymPy that currently don't work. With the downside being that the class introduced here could cause confusion in the future by people who are not as familiar with Python when it stops pretending it is a "traditional" programming language. ( I consider myself to be in that group to be honest)

oscarbenjamin commented:
I think that some of the benchmark results are accurate but some are spurious. The timings in CI are not completely consistent so a single run can't always be trusted but some of those changes show up regularly. These are the results from CI for the latest commit to master:
```diff
Run asv compare upstream/1.9 upstream/master --config asv.conf.actions.json --factor 1.5 --only-changed | tee master_vs_release_changed.txt
       before           after         ratio
     [907895ac]       [d822fcba]
                      <master>  
-         296±4ms          178±3ms     0.60  large_exprs.TimeLargeExpressionOperations.time_subs
-         281±3μs          134±2μs     0.48  matrices.TimeMatrixExpression.time_MatMul
-      18.4±0.2ms      10.1±0.04ms     0.55  matrices.TimeMatrixExpression.time_MatMul_doit
+     8.44±0.09ms       12.9±0.2ms     1.53  matrices.TimeMatrixPower.time_Case1
-      5.59±0.01s          435±4ms     0.08  polygon.PolygonArbitraryPoint.time_bench01
+     4.54±0.08ms       7.94±0.1ms     1.75  solve.TimeMatrixOperations.time_det(4, 2)
+     4.40±0.09ms       7.98±0.3ms     1.81  solve.TimeMatrixOperations.time_det_bareiss(4, 2)
+      43.1±0.7ms       64.7±0.3ms     1.50  solve.TimeMatrixPositiveDefinite.time_A_is_positive_definite
+        636±10ms         956±20ms     1.50  solve.TimeMatrixSolve.time_solve('GE')
+      49.1±0.4ms         95.1±2ms     1.94  solve.TimeMatrixSolvePyDySlow.time_linsolve(1)
+      50.1±0.7ms       94.0±0.6ms     1.88  solve.TimeMatrixSolvePyDySlow.time_solve(1)
```
Although the exact numbers differ those are the same benchmarks that we see highlighted in this PR. They aren't caused by the changes in this PR but they do represent an accumulation of slowdowns since 1.9.

ThePauliPrinciple commented:
I suspect this will need some documentation updates with `FuncHead` no longer printing as `type(self.obj)`. This should also make it clearer to the user what is going on, rather than pretending everything works as usual eventhough it doesn't.

ThePauliPrinciple commented:
Is it worth considering not to merge this before the upcomming release? So that this change can live in the development environment for a while? It's use is primarily for development in the library itself and should not be involved with users that much.

oscarbenjamin commented:
> Is it worth considering not to merge this before the upcomming release?

Yes, I think so.

ThePauliPrinciple commented:
> > Is it worth considering not to merge this before the upcomming release?
> 
> Yes, I think so.

If I understand correctly then this would now be merged in 1.11 dev.

Does anyone still have any concerns? I would love to get this merged since it is easy for non-conform classes to be introduced, although I of course understand it if there is more time needed for reviewing.

(Closing to rerun the checks.)

oscarbenjamin commented:
Looks good to me

oscargus commented:
There are some typos in the comments: `AtomExpr`, `AtomiExpr`, but I do not think it is a major issue so either fix those and merge or merge anyway. (Trusting @oscarbenjamin regarding mergability, the basic idea is indeed good!)

ThePauliPrinciple commented:
Is there a way to check if this has solved any test that is being skipped? I know that passing tests that are reported as failing are given as `xpassed`.

oscargus commented:
Not if they are explicitly skipped, no. Then they are never executed. We should probably only skip tests that are too time consuming to actually try (like the bicycle test or whichever it is that takes hours to run). Even if we know that the tests are currently failing it should be better to XFAIL them for this reason.

oscargus commented:
With that said, maybe there is a way to redefine the meaning of the skip decorator? But it will probably not help so much (unless one figures out a way to only run the skip-tests).

Now, we are not running a plain pytest, but this may work: https://stackoverflow.com/questions/55872637/how-to-run-a-test-marked-skip-in-pytest

oscarbenjamin commented:
> With that said, maybe there is a way to redefine the meaning of the skip decorator?

I would make decorators like `@needs_numpy`, `@needs_scipy` etc. Then there could be a way to make the tests not skip using e.g. an environment variable like `SYMPY_TEST_NOSKIP`. That would be useful for enforcing that no tests should be skipped in the optional dependency jobs in CI. If the decorator also applied a pytest marker like `needs_numpy` then you could do `pytest -m needs_numpy` to run only tests that need numpy.

The tests that are "too slow" should just be unconditionally skipped. Someone can always just import those tests directly and use them if they want but there's no ordinary circumstance where it makes sense to run those as part of a test run.

asmeurer commented:
You can run the slow tests with `./bin/test --slow`. However, be aware that this can take a long time to finish (I haven't tried it in a while, so I don't remember exactly how long, but it could be hours).

oscarbenjamin commented:
> You can run the slow tests with `./bin/test --slow`

There should be a separate option like `--too-slow` or something for running the ones that are skipped in CI. Otherwise they should always be skipped under both `bin/test` and `pytest`.

oscarbenjamin commented:
> it looks like if you run `./bin/test --slow` it does run tests that are otherwise skipped on CI

Yes, it does but it *shouldn't*. Using `--slow` should only run the tests that are run in the slow jobs in CI.

> IMO if a test is skipped unconditionally then there's little point to even having it at all.

The particularly slow tests can be run under a special flag like `bin/test --too-slow`. Also it should be possible to separate them with markers in pytest like `-m slow` or `-m too_slow`.

Basically it should be possible to clearly specify which tests you want to run rather than having the tests call skip internally.

ThePauliPrinciple committed:
core/basic: argument invariance for all Atom classes - class
The following identities now hold for `Atom` classes:
```python
assert obj.func == AtomSubClass
assert AtomSubClass == obj.func
assert hash(AtomSubClass) == hash(obj.func)
assert {AtomSubClass: True}[obj.func]
assert {obj.func: True}[AtomSubClass]
assert obj.func() == obj
assert obj.func(*obj.args) == obj
assert str(obj.func) == str(AtomSubClass)
```
These seem to all be needed since several parts of SymPy
assumes that `obj.func` returns `obj.__class__`, although it
is not needed for every class.

Several classes still behave as previously, these are:

`Quantity`
`BaseVector`
`BaseDyadic`
`BaseScalar`

It seems that these classes need a redesign rather than
a fix to their argument invariance, hence the previous
behaviour was maintained.
